### DISCUSSION TITLE
Themed Colors respecting colorMode

### DISCUSSION THREAD

**Discussion by amaster507** [chakra-ui/chakra-ui#2169] [Ideas]
# What I wanted to do: (From [Discord Discussion](https://discord.com/channels/660863154703695893/660865137137156146/762733474808594442))

Use Chakra themed colors by name and not need the useColorMode or useColorModeValue hooks to define my color switch logic in the components. If I could have a theme color named `background` and then that same color either be white or black based upon the colorMode that would be easier to use and less code in the components. Making the end use look like:

```html
<Box bg="background" color="text">Lorem Ipsum</Box>
```

In my theme file I can create a custom named color, but I cannot create a named color that respects the colorMode. What I tried was to pass a color using the mode function like:
```js
colors: {
  background: (props) => mode("#fff","#000")(props)
}
```

# Why it was not nice:

The theme color accepts either a static color or a object of shades having properties from 100(or 50?) to 900. The function defining the color based upon the props was never called.

# What I did instead:

Understanding that the colors is more or less a static property and wanting to still utilize the colorMode power of Chakra, I added an extra exportable object in my theme.js files to keep all of my colors together and then created a useColor hook that if used inside of the context wrapper of ChakraProvider will use the color based upon the colorMode very easily.

`theme.js:`
```js
const colors = {
  brand: "#eb8034",
  white: "#fefefe",
  black: "#8d8d8d",
  // ...
  ...theme.colors
}

export const themedColors = {
  ...colors,
  background: [color.white, color.black],
  text: [color.black, color.white]
}

export default extendTheme({
  // ...
  colors
})
```

`useColor.js`:
```js
import { useColorMode } from '@chakra-ui/core'
import { themedColors } from '../theme'

const useColor = (colorName) => {
  const { colorMode } = useColorMode()
  if (colorName in themedColors) {
    if (Array.isArray(themedColors[colorName]) && themedColors[colorName].length === 2) {
      return themedColors[colorName][(colorMode === "dark") ? 1 : 0]
    }
    return themedColors[colorName]
  }
  return null
}

export default useColor
```

Which I can then use like:

```js
<Box bg={useColor('background')} color={useColor('text')}>Lorem Ipsum</Box>
```

> **Note:** if you have a conditional rendering block you will need to use the hook higher in the component and set it to a variable using the variable in the conditional return statement.

**Comment by recursiveElk** [comment]
very nice write up, thanks for this!

**Comment by Xetera** [comment]
The way we got around this problem was by storing the color mode outside of chakra, generating multiple different themes for each color mode / theme and then re-rendering `<ChakraProvider theme={theme}>` with the correct one generated for the theme the user is on.

This way you can have components like

```js
<Text color="brand">...</Text>
```

that render a different color depending on the color mode without having to deal with hooks or generate multiple variants of `layerStyle`s.

I haven't seen any obvious drawbacks to using this approach so far. Until dynamic color definitions are added I think this is the only way to get chakra to play nicely with custom color palettes on websites with multiple color modes.

**Reply by adamtagscherer** [reply]
Hey guys, any update on this topic?
Does chakra support theme aware extended color values by now or do we have to make a workaround?
Will this feature available in the future?

Thanks!

**Reply by Xetera** [reply]
You still have to either use [`layerStyle`](https://chakra-ui.com/docs/features/text-and-layer-styles) or this approach if you want theme-specific colors.

**Comment by adamtagscherer** [comment]
Eventually I solved it another way which I share with you, it might be helpful for others.
```javascript
const extended = {
  styles: {
    global: (props) => ({
      ":root": {
        "--your-color": mode("#FF0000", "#00FF00")(props),
      }
    })
  }
}

const theme = extendTheme(extended);
```
You can use this color in your components like:
```jsx
<Box bgColor="var(--your-color)">Lorem ipsum</Box>
```
This way it will be color mode aware and you don't have to use useColorMode or useColorModeValue and it will be scalable.

**Comment by ryan-ju** [comment]
Sementic tokens might be what you need.  In short, they allow you to define tokens like `primary`, `secondary`, `success`, `error` that are theme aware.

https://chakra-ui.com/docs/features/semantic-tokens

**Comment by dherault** [comment]
Using semantic tokens, here is my solution:

```ts
const baseColors = {
  primary: {
    50: '#ffffff',
    100: '#eeeeee',
  }
}

export const lightColors = {
  ...baseColors,
  'text-primary': '#00030A',
  'background-primary': '#ffffff',
}

export const darkColors = {
  ...baseColors,
  'text-primary': '#ffffff',
  'background-primary': '#202229',
}

export const semanticColorTokens: Record<string, any> = {}

// Create semantic color tokens
Object.entries(lightColors).forEach(([colorKey, colorValue]) => {
  if (typeof colorValue === 'string') {
    semanticColorTokens[colorKey] = {
      default: colorValue,
      _dark: darkColors[colorKey as keyof typeof darkColors],
    }
  }
  else if (typeof colorValue === 'object') {
    Object.entries(colorValue).forEach(([shadeKey, shadeValue]) => {
      semanticColorTokens[`${colorKey}-${shadeKey}`] = {
        default: shadeValue,
        // @ts-expect-error
        _dark: darkColors[colorKey][shadeKey],
      }
    })
  }
})

const theme = extendTheme({
  ...baseTheme,
  colors: lightColors,
  semanticTokens: {
    colors: semanticColorTokens,
  },
})
```

and then:

```tsx
<Box bg="background-primary" />
```

Simply works in dark mode.
Hope this helps.
